helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -68,23 +71,47 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
-    // Sort the replicas so the input is stable for the greedy algorithm.
-    // For the other algorithm implementation, this sorting could be unnecessary.
-    for (AssignableReplica replica : getOrderedAssignableReplica(clusterModel)) {
+    List<AssignableReplica> toBeAssignedReplicas = clusterModel.getAssignableReplicaMap().values().stream()
+        .flatMap(replicas -> replicas.stream()).collect(Collectors.toList());
+    // Compute overall utilization of the cluster. Capacity dimension -> <total usage, total capacity>
+    Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap =
+        computeOverallClusterUtil(nodes);
+    Map<String, Integer> replicaHashCodeMap = toBeAssignedReplicas.parallelStream().collect(Collectors
```

Comment:
Please add some comments here since the logic has been moved away from where it is used. Might be hard to understand.